File: julia/discretized4dSphere.jl
using Quaternions 

type S3Grid
  depth::Int32
  vertices
  tetra
  tetra_levels
end

function InitS3Grid(s3::S3Grid)
  println("Initializing S3 Grid")
  # matrices are column major - so in order to be able to resize
  # efficiently we have one data-point per column
  s3.vertices = zeros(Float64, 4, 120)
  i = 1
  for a in [-1,1], b in [-1,1], c in [-1,1], d in [-1, 1]
    s3.vertices[:,i] = [a, b, c, d]
    i += 1
  end
  for j in 1:4, a in [-2, 2]
    s3.vertices[j,i] = a
    i+=1
  end
  # Golden Ratio
  phi = (1 + sqrt(5)) * 0.5
  # iterate over all *even* permutations
  # http://mathworld.wolfram.com/EvenPermutation.html
  for perm in [ (1,2,3,4), (1,3,4,2), (1,4,2,3), (2,1,4,3),
    (2,3,1,4), (2,4,3,1), (3,1,2,4), (3,2,4,1), (3,4,1,2),
    (4,1,3,2), (4,2,1,3), (4,3,2,1)]
    for a in [-1.,1.], b in [-1.,1.], c in [-1.,1.]
      s3.vertices[perm[1],i] = a*phi
      s3.vertices[perm[2],i] = b
      s3.vertices[perm[3],i] = c/phi
      s3.vertices[perm[4],i] = 0.
      i+=1
    end
  end
  s3.vertices *= 0.5
  # Tada: all of the vertices are unit length and hence \in S^3
  G = ones(Float64, 120, 120) * 99
  for i in 1:120, j in 1:120
    if j != i
      q1 = Quaternion(s3.vertices[1,i], s3.vertices[2,i],
        s3.vertices[3,i], s3.vertices[4,i])
      q2 = Quaternion(s3.vertices[1,j], s3.vertices[2,j],
        s3.vertices[3,j], s3.vertices[4,j])
      dq = normalize(q1 / q2)
      G[j,i] = angle(dq)
    end
  end
  minAngle = sort(G[:,1])[1]
  G[G .< minAngle - 1e-6] = -1
  G[G .> minAngle + 1e-6] = -1
  s3.tetra = zeros(Int32, 4, 600)
  i = 1
  for comb in combinations(1:120, 4) 
    if G[comb[1], comb[2]] > 0 && G[comb[1], comb[3]] > 0 && 
     G[comb[1], comb[4]] > 0 && G[comb[2], comb[3]] > 0 && 
     G[comb[2], comb[4]] > 0 && G[comb[3], comb[4]] > 0
      s3.tetra[:,i] = [comb[1], comb[2], comb[3], comb[4]]
      i += 1
    end
  end
  i == 601 && println("Wohoo found all the 600 Tetrahedra.")
  push!(s3.tetra_levels, 0)
  push!(s3.tetra_levels, 600)
end

function normalized(a)
  a /= sqrt(sum(a.^2))
end

function SubdivideOnce(s3::S3Grid)
  n_vertices = size(s3.vertices)[2]
  n_tetra = size(s3.tetra)[2]
  s3.vertices = reshape(resize!(s3.vertices[:], (n_vertices+n_tetra*6)*4), 
    4, n_vertices+n_tetra*6)
  s3.tetra = reshape(resize!(s3.tetra[:], (n_tetra*(8+1))*4),
    4, n_tetra*(8+1))
  push!(s3.tetra_levels, n_tetra * (8 + 1))
#  @sync @parallel for i in 1:n_tetra
  for i in 1:n_tetra
    i0 = s3.tetra[1,i]
    i1 = s3.tetra[2,i]
    i2 = s3.tetra[3,i]
    i3 = s3.tetra[4,i]
    i01 = n_vertices + (i-1)*6 + 1 
    i12 = n_vertices + (i-1)*6 + 2
    i20 = n_vertices + (i-1)*6 + 3
    i03 = n_vertices + (i-1)*6 + 4
    i13 = n_vertices + (i-1)*6 + 5
    i23 = n_vertices + (i-1)*6 + 6
    s3.vertices[:,i01] = normalized(s3.vertices[:,i0] + s3.vertices[:,i1])
    s3.vertices[:,i12] = normalized(s3.vertices[:,i1] + s3.vertices[:,i2]) 
    s3.vertices[:,i20] = normalized(s3.vertices[:,i2] + s3.vertices[:,i0]) 
    s3.vertices[:,i03] = normalized(s3.vertices[:,i0] + s3.vertices[:,i3]) 
    s3.vertices[:,i13] = normalized(s3.vertices[:,i1] + s3.vertices[:,i3]) 
    s3.vertices[:,i23] = normalized(s3.vertices[:,i2] + s3.vertices[:,i3]) 
    s3.tetra[:,n_tetra + (i-1)*8 + 1] = [i0, i01, i20, i03]
    s3.tetra[:,n_tetra + (i-1)*8 + 2] = [i01, i1, i12, i13]
    s3.tetra[:,n_tetra + (i-1)*8 + 3] = [i12, i2, i20, i23]
    s3.tetra[:,n_tetra + (i-1)*8 + 4] = [i03, i13, i23, i3]
    s3.tetra[:,n_tetra + (i-1)*8 + 5] = [i01, i13, i03, i20]
    s3.tetra[:,n_tetra + (i-1)*8 + 6] = [i01, i12, i13, i20]
    s3.tetra[:,n_tetra + (i-1)*8 + 7] = [i23, i20, i12, i13]
    s3.tetra[:,n_tetra + (i-1)*8 + 8] = [i23, i03, i20, i13]
  end
end

s3 = S3Grid(0, [], [], Int64[])
InitS3Grid(s3)
for lvl in 1:5
  println("Subdivide at level $lvl")
  SubdivideOnce(s3)
  println(s3.tetra_levels)
end

File: python/discretized4dSphere.py
import numpy as np
from js.geometry.rotations import *
from itertools import combinations, permutations

def normed(a):
  return a/np.sqrt((a**2).sum())

class Tetrahedron(object):
  def __init__(self, vertices, lvl):
    self.vertices = vertices
    self.tetra = np.arange(4)
    self.lvl = lvl
  def Center(self):
    center = 0.25*self.vertices[self.tetra[0]] \
      + 0.25*self.vertices[self.tetra[1]] + \
      0.25*self.vertices[self.tetra[2]] + 0.25*self.vertices[self.tetra[3]]
    return normed(center)
  def GeMinMaxVertexDotProduct(self):
    dots = np.zeros(6)
    for i, p in enumerate(combinations(range(4), 2)):
      print i,p, p[0], p[1]
      print self.vertices[p[0]]
      dots[i] = self.vertices[p[0]].dot(self.vertices[p[1]])
    return np.min(dots), np.max(dots)

  def Subdivide(self):
    '''
    Subdivide the tetraheadron 
    8 tetrahedra and pop out the inner corners of
    the new tetrahedra to the S3 sphere.
    http://www.ams.org/journals/mcom/1996-65-215/S0025-5718-96-00748-X/S0025-5718-96-00748-X.pdf
    '''
    tetrahedra = []
    vertices = np.zeros((6, 4))
    vertices[0, :] = normed(self.vertices[0] + self.vertices[1]) 
    vertices[1, :] = normed(self.vertices[1] + self.vertices[2]) 
    vertices[2, :] = normed(self.vertices[2] + self.vertices[0]) 
    vertices[3, :] = normed(self.vertices[0] + self.vertices[3]) 
    vertices[4, :] = normed(self.vertices[1] + self.vertices[3]) 
    vertices[5, :] = normed(self.vertices[2] + self.vertices[3]) 
    # Corner tetrahedron at 0th corner of parent.
    tetrahedra.append(Tetrahedron([
      self.vertices[0],
      vertices[0,:],
      vertices[2,:],
      vertices[3,:]], self.lvl+1))
    # Corner tetrahedron at 1th corner of parent.
    tetrahedra.append(Tetrahedron([
      self.vertices[1],
      vertices[0,:],
      vertices[1,:],
      vertices[4,:]], self.lvl+1))
    # Corner tetrahedron at 2th corner of parent.
    tetrahedra.append(Tetrahedron([
      self.vertices[2],
      vertices[1,:],
      vertices[2,:],
      vertices[5,:]], self.lvl+1))
    # Corner tetrahedron at 3th corner of parent.
    tetrahedra.append(Tetrahedron([
      self.vertices[3],
      vertices[3,:],
      vertices[4,:],
      vertices[5,:]], self.lvl+1))
    # Check which skew edge is the shortest.
    dots = np.zeros(3)
    dots[0] = vertices[0,:].dot(vertices[5,:])
    dots[1] = vertices[2,:].dot(vertices[4,:])
    dots[2] = vertices[3,:].dot(vertices[1,:])
    skewEdgeId = np.argmax(dots)
    if skewEdgeId == 0:
      tetrahedra.append(Tetrahedron([
        vertices[0,:], 
        vertices[5,:],
        vertices[3,:],
        vertices[2,:]], self.lvl+1))
      tetrahedra.append(Tetrahedron([
        vertices[0,:], 
        vertices[5,:],
        vertices[4,:],
        vertices[3,:]], self.lvl+1))
      tetrahedra.append(Tetrahedron([
        vertices[5,:], 
        vertices[0,:],
        vertices[1,:],
        vertices[4,:]], self.lvl+1))
      tetrahedra.append(Tetrahedron([
        vertices[5,:], 
        vertices[0,:],
        vertices[1,:],
        vertices[2,:]], self.lvl+1))
    elif skewEdgeId == 1:
      tetrahedra.append(Tetrahedron([
        vertices[0,:], 
        vertices[4,:],
        vertices[3,:],
        vertices[2,:]], self.lvl+1))
      tetrahedra.append(Tetrahedron([
        vertices[0,:], 
        vertices[1,:],
        vertices[4,:],
        vertices[2,:]], self.lvl+1))
      tetrahedra.append(Tetrahedron([
        vertices[5,:], 
        vertices[2,:],
        vertices[1,:],
        vertices[4,:]], self.lvl+1))
      tetrahedra.append(Tetrahedron([
        vertices[5,:], 
        vertices[3,:],
        vertices[2,:],
        vertices[4,:]], self.lvl+1))
    elif skewEdgeId == 2:
      tetrahedra.append(Tetrahedron([
        vertices[3,:], 
        vertices[1,:],
        vertices[0,:],
        vertices[2,:]], self.lvl+1))
      tetrahedra.append(Tetrahedron([
        vertices[3,:], 
        vertices[1,:],
        vertices[4,:],
        vertices[0,:]], self.lvl+1))
      tetrahedra.append(Tetrahedron([
        vertices[3,:], 
        vertices[1,:],
        vertices[5,:],
        vertices[4,:]], self.lvl+1))
      tetrahedra.append(Tetrahedron([
        vertices[1,:], 
        vertices[3,:],
        vertices[2,:],
        vertices[5,:]], self.lvl+1))
    return tetrahedra

class S3Grid(object):
  def __init__(self, depth):
    self.depth = depth
    self.vertices = np.zeros((120,4))
    self.tetra = np.zeros((600, 4), dtype = np.int)
    self.InitGrid()
    for lvl in range(self.depth):
      self.SubdivideOnce()
  def InitGrid(self):
    # The vertices of a 600-cell 
    vs = np.zeros((120,4))
    # https://en.wikipedia.org/wiki/600-cell
    # http://eusebeia.dyndns.org/4d/600-cell
    i=0
    for a in [-1.,1.]:
      for b in [-1.,1.]:
        for c in [-1.,1.]:
          for d in [-1.,1.]:
            vs[i,:] = [a,b,c,d]
            i+=1
    for j in range(4):
      vs[i,j] = 2.
      i+=1
      vs[i,j] = -2.
      i+=1
    # Golden Ratio
    phi = (1 + np.sqrt(5)) * 0.5
    # iterate over all *even* permutations
    # http://mathworld.wolfram.com/EvenPermutation.html
    for perm in [ [0,1,2,3], [0,2,3,1], [0,3,1,2], [1,0,3,2],
        [1,2,0,3], [1,3,2,0], [2,0,1,3], [2,1,3,0], [2,3,0,1],
        [3,0,2,1], [3,1,0,2], [3,2,1,0]]:
      for a in [-1.,1.]:
        for b in [-1.,1.]:
          for c in [-1.,1.]:
            vs[i,perm[0]] = a*phi
            vs[i,perm[1]] = b
            vs[i,perm[2]] = c/phi
            vs[i,perm[3]] = 0.
            i+=1
    vs *= 0.5
    # Filter out half the sphere
    north = np.array([1.,0.,0.,0.])
    j = 0
    for i in range(vs.shape[0]):
      if np.arccos(vs[i,:].dot(north)) <= 120*np.pi/180.:
        vs[j,:] = vs[i,:]
        j+=1
    self.vertices = np.copy(vs[:j,:])
    n_vertices = self.vertices.shape[0]
    print self.vertices.shape
      
    # Tada: all of the vertices are unit length and hence \in S^3
#    print np.sqrt((self.vertices**2).sum(axis=1))
    if False:
      import mayavi.mlab as mlab
      figm = mlab.figure()
      for i in range(n_vertices):
        qi = Quaternion(vec=self.vertices[i])
        qi.plot(figm)
      mlab.show(stop=True)

    G = np.ones((n_vertices, n_vertices)) * 99
    for i in range(n_vertices):
      for j in range(n_vertices):
        if j != i:
          qi = Quaternion(vec=self.vertices[i])
          qj = Quaternion(vec=self.vertices[j])
          G[i,j] = qi.angleTo(qj)

    GSorted = np.sort(G, axis=1)
#    print GSorted[0,:] * 180. / np.pi

    minAngle = GSorted[0, 1] 
    G[G < minAngle - 1e-6] = -1.
    G[G > minAngle + 1e-6] = -1.

    if False:
      import matplotlib.pyplot as plt
      plt.figure()
      plt.imshow(G, interpolation="nearest")
      plt.show()

    tetra = []
    for p in combinations(range(n_vertices), 4):
      if G[p[0], p[1]] > 0 and G[p[0], p[2]] > 0 and G[p[0], p[3]] > 0 and  \
        G[p[1], p[2]] > 0 and G[p[1], p[3]] > 0 and G[p[2], p[3]] > 0:
        tetra.append(p)
    print len(tetra)
#    if len(tetra) == 600:
#      print "Yeaha the 600 cell"
    self.tetra_levels = [0, len(tetra)]
    self.tetra = np.array(tetra)
          

  def SubdivideOnce(self):
      '''
      Subdivide each of the existing tetraheadra of the current
      S3Grid into 4 triangles and pop out the inner corners of
      the new triangles to the S3 sphere.
      http://www.ams.org/journals/mcom/1996-65-215/S0025-5718-96-00748-X/S0025-5718-96-00748-X.pdf
      '''
      n_vertices = self.vertices.shape[0]
      n_tetra = self.tetra.shape[0]
      print n_tetra, n_vertices
      self.vertices.resize((n_vertices + n_tetra * 6, 4), refcheck=False)
      self.tetra.resize((n_tetra * (8 + 1), 4))
      self.tetra_levels.append(n_tetra * (8 + 1))
      for i in range(n_tetra):
          i0 = self.tetra[i, 0]
          i1 = self.tetra[i, 1]
          i2 = self.tetra[i, 2]
          i3 = self.tetra[i, 3]
          i01 = n_vertices + i*6 
          i12 = n_vertices + i*6 + 1
          i20 = n_vertices + i*6 + 2
          i03 = n_vertices + i*6 + 3
          i13 = n_vertices + i*6 + 4
          i23 = n_vertices + i*6 + 5
          self.vertices[i01, :] = 0.5 * (self.vertices[i0, :] + self.vertices[i1, :]) 
          self.vertices[i12, :] = 0.5 * (self.vertices[i1, :] + self.vertices[i2, :]) 
          self.vertices[i20, :] = 0.5 * (self.vertices[i2, :] + self.vertices[i0, :]) 
          self.vertices[i03, :] = 0.5 * (self.vertices[i0, :] + self.vertices[i3, :]) 
          self.vertices[i13, :] = 0.5 * (self.vertices[i1, :] + self.vertices[i3, :]) 
          self.vertices[i23, :] = 0.5 * (self.vertices[i2, :] + self.vertices[i3, :]) 
          self.tetra[n_tetra + i*8, :]     = [i0, i01, i20, i03]
          self.tetra[n_tetra + i*8 + 1, :] = [i01, i1, i12, i13]
          self.tetra[n_tetra + i*8 + 2, :] = [i12, i2, i20, i23]
          self.tetra[n_tetra + i*8 + 3, :] = [i03, i13, i23, i3]
          self.tetra[n_tetra + i*8 + 4, :] = [i01, i13, i03, i20]
          self.tetra[n_tetra + i*8 + 5, :] = [i01, i12, i13, i20]
          self.tetra[n_tetra + i*8 + 6, :] = [i23, i20, i12, i13]
          self.tetra[n_tetra + i*8 + 7, :] = [i23, i03, i20, i13]
      print np.sqrt((self.vertices[n_vertices::,:]**2).sum(axis=1))
      self.vertices[n_vertices::, :] /= np.sqrt((self.vertices[n_vertices::,:]**2).sum(axis=1))[:, np.newaxis]
  def GetTetras(self, level):
    return self.tetra[self.tetra_levels[level]: \
        self.tetra_levels[level+1]]
  def GetVertex(self, id):
    return self.vertices[id,:]
  def GetTetrahedra(self, level):
    tetras = self.GetTetras(level)
    tetrahedra = []
    for tetra in tetras:
      tetrahedra.append(Tetrahedron([
        self.vertices[tetra[0],:],
        self.vertices[tetra[1],:],
        self.vertices[tetra[2],:],
        self.vertices[tetra[3],:]
        ], level))
    return tetrahedra

if __name__ == "__main__":
  lvls = 4
  s3 = S3Grid(lvls)
  print s3.tetra_levels

  for lvl in range(lvls):
    print " --- {} ---".format(lvl)
    tetras = s3.GetTetras(lvl)
    dAng = np.zeros((tetras.shape[0], 6))
    for i, tetra in enumerate(tetras):
      dAng[i, 0] = Quaternion(vec = s3.GetVertex(tetra[0])).angleTo(
              Quaternion(vec = s3.GetVertex(tetra[1])))
      dAng[i, 1] = Quaternion(vec = s3.GetVertex(tetra[1])).angleTo(
              Quaternion(vec = s3.GetVertex(tetra[2])))
      dAng[i, 2] = Quaternion(vec = s3.GetVertex(tetra[0])).angleTo(
              Quaternion(vec = s3.GetVertex(tetra[2])))
      dAng[i, 3] = Quaternion(vec = s3.GetVertex(tetra[0])).angleTo(
              Quaternion(vec = s3.GetVertex(tetra[3])))
      dAng[i, 4] = Quaternion(vec = s3.GetVertex(tetra[1])).angleTo(
              Quaternion(vec = s3.GetVertex(tetra[3])))
      dAng[i, 5] = Quaternion(vec = s3.GetVertex(tetra[2])).angleTo(
              Quaternion(vec = s3.GetVertex(tetra[3])))
    print dAng.mean(axis=0) * 180. / np.pi
